tests/subsys/tracing/src/main.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <ztest.h>
+#if defined(CONFIG_TRACING_BACKEND_UART) || defined(CONFIG_TRACING_BACKEND_USB)
+#include "../../../../subsys/tracing/include/tracing_backend.h"
+#endif
+
+/**
+ * @defgroup subsys_tracing_tests subsys tracing
+ * @ingroup all_tests
+ * @{
+ * @}
+ */
+
+/* size of stack area used by each thread */
+#define STACK_SIZE (512 + CONFIG_TEST_EXTRA_STACKSIZE)
+#define LOOP_TIMES 10
+
+bool thread_switched_found;
+bool sys_trace_void_found;
+bool sys_trace_semaphore_found;
+
+static struct k_thread thread;
+K_THREAD_STACK_DEFINE(thread_stack, STACK_SIZE);
+
+/* define 2 semaphores */
+K_SEM_DEFINE(thread1_sem, 1, 1);
+K_SEM_DEFINE(thread2_sem, 0, 1);
+
+/* thread handle for switch */
+void thread_handle(void *p1, void *self_sem, void *other_sem)
+{
+	for (int i = 0; i < LOOP_TIMES; i++) {
+		/* take my semaphore */
+		k_sem_take(self_sem, K_FOREVER);
+		/* wait for a while, then let other thread have a turn. */
+		k_sleep(K_MSEC(100));
+
+		k_sem_give(other_sem);
+	}
+}
+
+#if (defined(CONFIG_TRACING_BACKEND_UART) || defined(CONFIG_TRACING_BACKEND_USB))
+static void tracing_backends_output(
+		const struct tracing_backend *backend,
+		uint8_t *data, uint32_t length)
+{
+	uint8_t recv_char[100], recv_index = 0;
```

Comment:
I don't think you have to define a new buffer. Only using "*data" is enough.

Reply:
recv_char[] is used to store a whole line string,  and *data just get a char from backend,  so we need define a new buffer here.

Reply:
Let me have a try,